### ISSUE TITLE
checksum problem

### ISSUE THREAD

**Issue by ritchieGitHub** [nom-tam-fits/nom-tam-fits#26]
an additional test case for checksum showed a strange behavior with non constant checksums:

nom.tam.fits.test.ChecksumTest.testCheckSum2()

test is ignored for now.

**Comment by ritchieGitHub** [comment]
because in the comming fits standard this is fixated in more detail the fix will be done during the implementation of the upcomming standard of fits

**Comment by tony-johnson** [comment]
There is actually a serious bug in the existing FitsCheckSum class, which may or may not be related to the test failure described above, but certainly prevents it from being used to generate correct checksums. The fix is:

```java
@@ -113,8 +113,8 @@
         long hicarry = hi >>> FitsIO.BITS_OF_2_BYTES;
         long locarry = lo >>> FitsIO.BITS_OF_2_BYTES;
         while (hicarry != 0 || locarry != 0) {
-            hi = (hi & FitsIO.SHORT_OF_LONG_MASK) | locarry;
-            lo = (lo & FitsIO.SHORT_OF_LONG_MASK) | hicarry;
+            hi = (hi & FitsIO.SHORT_OF_LONG_MASK) + locarry;
+            lo = (lo & FitsIO.SHORT_OF_LONG_MASK) + hicarry;
             hicarry = hi >>> FitsIO.BITS_OF_2_BYTES;
             locarry = lo >>> FitsIO.BITS_OF_2_BYTES;
         }
```

**Comment by ritchieGitHub** [comment]
That's great thanks for the solution, are you sure it is correct now? please test the 1.14.0-SNAPSHOT version. should be availabe in 20 minutes

**Comment by ritchieGitHub** [comment]
Any progress in the tests?

**Comment by tony-johnson** [comment]
Hi, yes the fix committed to 1.40.0-SNAPSHOT fixed the issue I reported with the checksums. 

There is a more subtle issue, which is probably not amenable to a quick fix. The existing FitsCheckSum class works by "writing" the contents of the header to an in memory buffer, then calculating the checksum on that buffer. There are two issues with this, one is that you have a big HDU it can use a lot of extra memory, and second it has an undesirable side effect of changing the internal state of the HDU, due to the way that elements like the Header class cache items like fileOffset. So while it should be possible to read an HDU from a file, update the checksum and rewrite() the HDU, this is in fact not possible because of the side effect of calculating the checksum.

For this reasons and others for the time being I rewrote the FitsCheckSum class to handle my own requirements. I have put the updated class here:  

https://gist.github.com/tony-johnson/a88e896fda979d33bf2d

in case any parts may be of more general use.
